subsys/bluetooth/mesh/pb_adv.c

```diff
@@ -317,6 +359,142 @@ static void protocol_timeout(struct k_work *work)
 	link.rx.seg = 0U;
 	prov_link_close(PROV_BEARER_LINK_STATUS_TIMEOUT);
 }
+
+static struct delayable_adv_ctx *delayable_adv_ctx_get_unused(void)
+{
+	for (int i = 0; i < CONFIG_BT_MESH_PB_ADV_DELAYABLE_ADV_COUNT; i++) {
+		if (!atomic_test_bit(delayable_advs_ctx[i].flags, DELAYABLE_ADV_CTX_IN_USE)) {
+			return &delayable_advs_ctx[i];
+		}
+	}
+
+	return NULL;
+}
+
+static struct delayable_adv_ctx *delayable_adv_ctx_assign(struct bt_mesh_adv *adv,
+							  const struct bt_mesh_send_cb *cb,
+							  void *cb_data, bool retransmit)
+{
+	struct delayable_adv_ctx *ctx;
+
+	ctx = delayable_adv_ctx_get_unused();
+	if (!ctx) {
+		LOG_WRN("Purge pending delayed advertisement.");
+		if (!delayable_adv_ctx_push()) {
+			return NULL;
+		}
+
+		ctx = delayable_adv_ctx_get_unused();
+		if (!ctx) {
+			return NULL;
+		}
+	}
+
+	atomic_set_bit(ctx->flags, DELAYABLE_ADV_CTX_IN_USE);
+	ctx->adv = adv;
+	ctx->cb = cb;
+	ctx->cb_data = cb_data;
+	if (retransmit) {
+		atomic_set_bit(ctx->flags, DELAYABLE_ADV_CTX_RETRANSMIT);
+	}
+
+	return ctx;
+}
+
+static void delayable_adv_ctx_release(struct delayable_adv_ctx *ctx)
+{
+	if (!atomic_test_and_clear_bit(ctx->flags, DELAYABLE_ADV_CTX_RETRANSMIT)) {
+		bt_mesh_adv_unref(ctx->adv);
+	}
+
+	(void)sys_slist_find_and_remove(&link.delayable_advs, &ctx->node);
+	atomic_clear_bit(ctx->flags, DELAYABLE_ADV_CTX_IN_USE);
+
+}
+
+static struct delayable_adv_ctx *delayable_adv_ctx_head_get(void)
+{
+	sys_snode_t *node = sys_slist_peek_head(&link.delayable_advs);
+
+	if (!node) {
+		return NULL;
+	}
+
+	return CONTAINER_OF(node, struct delayable_adv_ctx, node);
+}
+
+static void delayable_adv_ctx_reschedule(void)
+{
+	uint16_t random_delay;
+	k_timeout_t delay = K_NO_WAIT;
+
+	if (sys_slist_is_empty(&link.delayable_advs)) {
+		return;
+	}
+
+	(void)bt_rand(&random_delay, sizeof(random_delay));
+	random_delay = 20 + (random_delay % 30);
+	delay = K_MSEC(random_delay);
+
+	LOG_DBG("delay %u", random_delay);
+
+	k_work_reschedule(&link.random_delay, delay);
+}
+
+static bool delayable_adv_ctx_push(void)
+{
+	struct delayable_adv_ctx *ctx;
+
+	ctx = delayable_adv_ctx_head_get();
+
+	if (!ctx) {
+		return false;
+	}
+
+	LOG_DBG("%u bytes: %s", ctx->adv->b.len, bt_hex(ctx->adv->b.data, ctx->adv->b.len));
+	bt_mesh_adv_send(ctx->adv, ctx->cb, ctx->cb_data);
+
+	delayable_adv_ctx_release(ctx);
+	return true;
+}
+
+static void delayable_adv_handler(struct k_work *work)
+{
+	delayable_adv_ctx_push();
+	delayable_adv_ctx_reschedule();
+}
+
+static void adv_delay_send(struct bt_mesh_adv *adv, const struct bt_mesh_send_cb *cb,
+			   void *cb_data, bool retransmit)
+{
+	struct delayable_adv_ctx *ctx;
+
+	ctx = delayable_adv_ctx_assign(adv, cb, cb_data, retransmit);
+
+	if (!ctx) {
+		LOG_WRN("Unable to allocate delayable advertiser context, sending immediately.");
```

Comment:
This brakes the shall statement of the spec regarding random delay. Can we allocate enough buffers to avoid this situation?

Reply:
This was a solution to work around, this comment "https://github.com/zephyrproject-rtos/zephyr/pull/65629#discussion_r1439282764". What is best? Just drop the message and resend everything if retransmit is set, ack will not be sent if this happens.